File: Stream.py
from Packet import Packet
from Demultiplexor import Demultiplexor

PACKET_SIZE = 188


class Stream:
    filename = ""

    def __init__(self, filename):
        self.filename = filename
        self.in_file = self.open_file()
        self.out_file = self.open_file(out=True)
        self.demultiplexor = Demultiplexor()

    def open_file(self, out=False):
        if not out:
            try:
                return open(self.filename, 'rb')
            except OSError as e:
                print("ERROR: File \"" + self.filename + "\" cannot be opened.")
                exit(2)
        filename = self.filename.replace('.ts', '.txt')
        return open(filename, 'w')

    def parse_file(self):
        while True:
            buffer = self.in_file.read(PACKET_SIZE)
            if buffer == '':
                break
            elif len(buffer) != PACKET_SIZE:
                break

            packet = Packet(PACKET_SIZE, buffer)
            bad_sync_byte = packet.parse()

            if not bad_sync_byte:
                continue

            self.demultiplexor.packet_num += 1

            # If there is error in the packet or pid is 0x1fff => skip this packet
            if packet.transport_error_ind or packet.pid == 0x1fff:
                continue
            # packet.print_header()

            if packet.pid == self.demultiplexor.PID_PAT and not self.demultiplexor.PAT_analysed:
                self.demultiplexor.parse_PAT(packet)
                # break
            elif packet.pid == self.demultiplexor.PID_NIT and not self.demultiplexor.NIT_analysed:
                self.demultiplexor.parse_NIT(packet)
                # break
            elif packet.pid == self.demultiplexor.PID_SDT:
                self.demultiplexor.parse_SDT(packet)
                # break
        # TODO: add

File: c++/Stream.cpp
//
// Created by odehnaltomas on 24.10.19.
//

#include "Stream.h"
#include "Packet.h"
#include "Demultiplexor.h"

Stream::Stream(const char *filename) {
	Stream::filename = string(filename);
	open_file();
};

bool Stream::open_file() {
	input_file.open(filename, ios::binary);

	uint8_t i = filename.find_last_of(".");
	output_file.open(filename.substr(0, i).append(".txt"), ios::out);
};

void Stream::parse_file() {
	while(true) {
		input_file.read(reinterpret_cast<char*>(buffer), PACKET_SIZE);
		if(input_file.gcount() != PACKET_SIZE) {
			break;
		}

		Packet packet = Packet(buffer, PACKET_SIZE);
		bool bad_sync_byte = packet.parse_header();

		demultiplexor.increment_packet_numb();
		demultiplexor.increment_packet_num_by_pid(packet.get_pid());

		if(!bad_sync_byte) {
			continue;
		}
		if(packet.get_transport_err_ind() || packet.get_pid() == 0x1FFF) {
			continue;
		}


		if(packet.get_pid() == demultiplexor.PID_PAT && !demultiplexor.is_pat_analysed()) {
			demultiplexor.parse_pat(packet);
		}
		else if(packet.get_pid() == demultiplexor.PID_NIT && !demultiplexor.is_nit_analysed()) {
			demultiplexor.parse_nit(packet);
		}
		else if(packet.get_pid() == demultiplexor.PID_SDT && !demultiplexor.is_sdt_analysed()) {
			demultiplexor.parse_sdt(packet);
		}
		else if(demultiplexor.is_pat_analysed() && demultiplexor.check_pmt(packet.get_pid())) {
			demultiplexor.parse_pmt(packet);
		}

	}
//	demultiplexor.print_pat_data();
//	demultiplexor.print_nit_data();
//	demultiplexor.print_sdt_data();
	demultiplexor.calculate_bitrate();
	demultiplexor.bind_pids_and_counters();
	demultiplexor.save_data_to_file(output_file);
};